### ISSUE TITLE
Possibility to ignore a single diagnostic

### ISSUE THREAD

**Issue by bellini666** [microsoft/pyright#2877]
Hello,

I really miss sometimes the ability to `# type:ignore` a specific diagnostic issue. The reasoning for this is that, even though I have sometimes to ignore a line for some specific reason, I don't want everything to be ignored there, which might mask an issue in the future.

This could be something similar to [`# noqa` for flake8](https://flake8.pycqa.org/en/3.1.1/user/ignoring-errors.html#in-line-ignoring-errors) where this simple form ignores everything, but you can ` # noqa: FOO` to ignore `FOO` in that line/context or even `# noqa: FOO,BAR` to ignore both `FOO` and `BAR`.

Sorry if this is already supported or has already been discussed, I tried searching in the documentation and issues but didn't find anything.

**Comment by erictraut** [comment]
This has already been discussed and (so far) rejected.

The `# type: ignore` is a standard mechanism defined in PEP 484 for suppressing diagnostics on a line, and pyright supports it. In general, I think that `# type: ignore` is a bad idea and should never be used in a code base. It's too blunt of an instrument. It's effectively a landmine that will eventually blow up in someone's face. There's almost always a better solution than using `# type: ignore`. Pyright even has an option that allows all `# type: ignore` comments to be disabled, and I use that in my team's Python code base. It's a large code base, and we've found that living without `# type: ignore` is not only possible but also preferable. I would encourage others to take this approach as well if you're serious about static type checking.

I recently added support for `reportUnnecessaryTypeIgnoreComment` that makes `# type: ignore` slightly less terrible. It allows you to know when a `# type: ignore` is no longer necessary.

Diagnostic category names are not standardized across type checkers, so there's no standard way to refer to these categories across mypy, pyright, etc.

Pyright already provides the ability to enable or disable specific pyright-specific diagnostic categories on a per-file basis through the use of comments of the form `# pyright: reportUnnecessaryTypeIgnoreComment=true` or `# pyright: strict`.

For all of the reasons above, I don't think it makes sense to add further support for suppressing specific diagnostic categories on a line.

**Comment by erictraut** [comment]
We might revisit this in the future, but for now, I'm going to close.

**Comment by bellini666** [comment]
Hey @erictraut ,

I know that you already closed the issue, but I just want to say that the reason I asked for this is exactly because of all of what you stated.

Specifically because of, quoting you, `It's too blunt of an instrument. It's effectively a landmine that will eventually blow up in someone's face.`, I want to ignore some specific errors but not all. I can agree that there's always a better way when it comes to my code or my project's code, but there's a lot of poor typed (or even wrongly typed) libs that I have to `# type:ignore` for some specific reason. In this specific case I would love to ignore that specific issue but still be alerted of something else in the future to avoid that it "blow up in my face".

The same logic applies to the per-file bases ignore, but in an even worse scenario IMO.

In summary, for me it is like the `try: except` argument that: You want the minimum amount of code inside it, and also you want to catch the most specific exceptions possible to avoid letting some real error pass silently.

Well, this is just my 2 cents here. Hope this gets revisited in the future, but I also understand your point of view here.

**Comment by erictraut** [comment]
> but there's a lot of poor typed (or even wrongly typed) libs

IMO, that's not a good excuse to use `# type: ignore`. You can create your own minimal type stubs for libraries that are completely untyped, and you can use `cast` to work around incorrect types or other type annotation issues. My team has a 500K Python code base, and it doesn't contain any `# type: ignore` comments.

I'll point out that it's highly unlikely that a line that has suppressed diagnostics would suddenly develop a diagnostic from a completely different class. What's _much_ more likely is that the error you suppressed masks an actual problem that later manifests as an unintended runtime error.

**Comment by Fmajor** [comment]
Is there a way to suppress errors like '"xxx" is not defined' for a single file?
the `# type: ignore` can only suppress error of a single line, I want to suppress all errors with the same error type and error msg

**Comment by erictraut** [comment]
@Fmajor, can you provide an example of where a variable is defined in a way that pyright cannot detect? In most of those cases, you can modify the code in a way that makes the variable visible to a static analyzer. If that's not possible, you could suppress the `reportUndefinedVariable` diagnostic rule for an entire file.

**Comment by Fmajor** [comment]
when you use `exec(dynamic_code, globals(), locals())`
how can i suppress the `reportUndefinedVariable` for the entire file?

**Comment by erictraut** [comment]
You can add the following comment somewhere in the file:
```python
# pyright: reportUndefinedVariable=false
```

For more details, refer to [this documentation](https://github.com/microsoft/pyright/blob/main/docs/comments.md).

Alternatively, you can initialize the variables in a conditional `TYPE_CHECKING` block.
```python
from typing import TYPE_CHECKING
if TYPE_CHECKING:
    my_global1 = []
    my_global2 = 0
    ...
```

**Comment by Fmajor** [comment]
Thanks!

**Comment by erictraut** [comment]
After much discussion, we've decided to add support for a `# pyright: ignore` comment. This also supports rule-specific suppression, as in `# pyright: ignore [reportUnusedImport, reportGeneralTypeIssues]`. For more details, refer to [this documentation](https://github.com/microsoft/pyright/blob/main/docs/comments.md#line-level-diagnostic-suppression). This will be included in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.229, which I just published. It will also be included in the next release of pylance.
